### ISSUE TITLE
Registry unnecessarily redirects blob HEAD requests to storage URL

### ISSUE THREAD

**Issue by aaronlehmann** [distribution/distribution#3691]
When the "redirect" option is set to true, it applies to both `HEAD` and `GET` requests for blobs. The use case for `GET`s makes sense - this avoids the registry needing to proxy potentially large blobs from the storage backend to the client. But in the case of `HEAD` requests, the registry [has already done a `Stat`](https://github.com/distribution/distribution/blob/8857a1948739db3520e2fb5211979aa62c3dd5da/registry/storage/blobserver.go#L27), and could simply return that information to the client, but instead is issuing a redirect which can add considerable latency.

Am I missing anything that justifies this behavior? Would you be open to a PR to change `ServeBlob` to avoid redirecting `HEAD` requests?

**Comment by aaronlehmann** [comment]
@milosgajdos: Any thoughts on this?

**Comment by thaJeztah** [comment]
From your description, the current behavior seems to comply with the HTTP specs; from [RFC9110](https://www.rfc-editor.org/rfc/rfc9110.html#name-head)

> [9.3.2. ](https://www.rfc-editor.org/rfc/rfc9110.html#section-9.3.2)[HEAD](https://www.rfc-editor.org/rfc/rfc9110.html#name-head)
>
> The HEAD method is identical to GET except that the server MUST NOT send content in the response. HEAD is used to obtain metadata about the [selected representation](https://www.rfc-editor.org/rfc/rfc9110.html#selected.representation) without transferring its representation data, often for the sake of testing hypertext links or finding recent modifications.
> The server SHOULD send the same header fields in response to a HEAD request as it would have sent if the request method had been GET. However, a server MAY omit header fields for which a value is determined only while generating the content. For example, some servers buffer a dynamic response to GET until a minimum amount of data is generated so that they can more efficiently delimit small responses or make late decisions with regard to content selection. Such a response to GET might contain [Content-Length](https://www.rfc-editor.org/rfc/rfc9110.html#field.content-length) and [Vary](https://www.rfc-editor.org/rfc/rfc9110.html#field.vary) fields, for example, that are not generated within a HEAD response. These minor inconsistencies are considered preferable to generating and discarding the content for a HEAD request, since HEAD is usually requested for the sake of efficiency.


I guess this part may be relevant here;

> However, a server MAY omit header fields for which a value is determined only while generating the content. 

Performing the `Stat` would fall within that bit, but I guess it still needs to do the Stat _in order to_ get the redirect location?

**Comment by aaronlehmann** [comment]
I don't think it would violate the RFC for the registry to generate its own response to a `HEAD` using the stat information it already has. In fact, the RFC seems to encourage this kind of optimization:

> These minor inconsistencies are considered preferable to generating and discarding the content for a HEAD request, since HEAD is usually requested for the sake of efficiency.

**Comment by thaJeztah** [comment]
Thinking a bit further; so, in your case, there's `HEAD` request, but the client sees the `redirect`, so follows it?

Wondering if that's the part to look into; basically; if the `HEAD` has the headers it need, it wouldn't have to follow the redirect and do another request. I think that would still comply to the RFC if we do the same for both `GET` and `HEAD`

The cache headers should probably be omitted though in that case, as (IIUC) the redirect-to URL would be an ephemeral URL, so result should not be cached? https://github.com/distribution/distribution/blob/8857a1948739db3520e2fb5211979aa62c3dd5da/registry/storage/blobserver.go#L59-L60

**Comment by thaJeztah** [comment]
> I don't think it would violate the RFC for the registry to generate its own response to a HEAD using the stat information

Right, so I think there's two bits here;

1. Add the extra headers https://github.com/distribution/distribution/blob/8857a1948739db3520e2fb5211979aa62c3dd5da/registry/storage/blobserver.go#L62-L74


2. Remove the `Location` header


From the above, it feels like `1.` is fine, but `2.` may not be.

However, I think clients should be allowed to ignore a `Location` header (don't follow redirects), so in case the response contains the headers they need, they can ignore the `Location` header and be done. Clients that don't have that logic would continue to follow the redirect.

**Comment by aaronlehmann** [comment]
> However, I think clients should be allowed to ignore a `Location` header (don't follow redirects), so in case the response contains the headers they need, they can ignore the `Location` header and be done. Clients that don't have that logic would continue to follow the redirect.

I'm pretty sure existing clients do not have htis logic. Anything written in Go (such as Docker) will follow redirects automatically by default, unless this is explicitly disabled. I do not see the `distribution` client disabling redirect following: https://github.com/distribution/distribution/blob/8857a1948739db3520e2fb5211979aa62c3dd5da/registry/client/repository.go#L876-L880

Anyway, I think it would make a lot of sense to at least provide an opt-in setting for people who prefer efficiency and cost savings over strict RFC compliance.
